cassandra/cluster.py

```diff
@@ -3497,6 +3497,10 @@ def _cancel_timer(self):
 
     def _on_timeout(self):
 
+        # PYTHON-853: for short timeouts, we sometimes race with our __init__
+        if self._connection is None:
+            self._timer = self.session.cluster.connection_class.create_timer(0.01, self._on_timeout)
+
```

Comment:
Should there be a return here?

Reply:
Whoops, yes

Reply:
I like this patch since it's really a solution. But I think it can get into a loop sometimes, for example, against only one C* node:
```python
import time
import logging

log = logging.getLogger()
log.setLevel('DEBUG')
handler = logging.StreamHandler()
handler.setFormatter(logging.Formatter("%(asctime)s [%(levelname)s] %(name)s: %(message)s"))
log.addHandler(handler)

from cassandra.cluster import Cluster

cluster = Cluster()
session = cluster.connect()
print(cluster.metadata.all_hosts())

assert len(session._pools.values()) == 1
tuple(session._pools.values())[0].shutdown()

# stuck here
session.execute("SELECT * from system.local", timeout=0)
```
Shutting down the pool `tuple(session._pools.values())[0].shutdown()` like that is not very legit but it *could* happen before some execute. In this case `self._connection == None` and is never set, therefore `self._on_timeout()` is in a loop calling itself.